### PR TITLE
chore: set initial host list provider according to the dialect

### PR DESCRIPTION
**Pull request by karenc-bq**
### Summary

The current implementation sets the initial host list provider to ConnectionStringHostListProvider regardless of the dialect. User applications need to ensure AuroraHostListProvider is set through the AuroraHostListConnectionPlugin if the intended plugin doesn't automatically updates the host list provider.

This change automatically sets the AuroraHostListProvider for cluster-aware dialects and deprecates the need for AuroraHostListConnectionPlugin.

### By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: wrapper/src/main/java/software/amazon/jdbc/wrapper/ConnectionWrapper.java [hunk 1]
```diff
@@ -106,7 +111,8 @@ protected ConnectionWrapper(
       throw new IllegalArgumentException("url");
     }
 
-    init(props, connectionPluginManager, pluginService, hostListProviderService, pluginManagerService);
+    init(props, connectionPluginManager, pluginService, hostListProviderService,
```
Comment (brunos-bq) [inline wrapper/src/main/java/software/amazon/jdbc/wrapper/ConnectionWrapper.java:114]: NIT: are those changes based on a checkstyle? it didn't seem that the line was that long to require a break

File: CHANGELOG.md [hunk 2]
```diff
@@ -3,6 +3,10 @@ All notable changes to this project will be documented in this file.
 
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and this project adheres to [Semantic Versioning](https://semver.org/#semantic-versioning-200).
 
+## [2.2.x] - ???
+### :crab: Changed
+- Dynamically set the default host list provider based on the dialect.
```
Comment (brunos-bq) [inline CHANGELOG.md:8]: Not sure if i got it entirely, but what about:

"Dynamically sets the default hostlist provider based on the dialect used. AuroraHostList plugin is not required anymore when using TopologyAwareDatabaseCluster dialect."

File: CHANGELOG.md [hunk 3]
```diff
@@ -3,6 +3,10 @@ All notable changes to this project will be documented in this file.
...
+
```
Comment (brunos-bq) [inline CHANGELOG.md:9]: Add a line to mention that the AuroraHostListPlugin was removed.

File: docs/using-the-jdbc-driver/UsingTheJdbcDriver.md [hunk 4]
```diff
@@ -108,7 +108,6 @@ The AWS JDBC Driver has several built-in plugins that are available to use. Plea
...
-| Aurora Host List Connection Plugin                                                             | `auroraHostList`          | Aurora                 | Retrieves Amazon Aurora clusters information. <br><br>**:warning:Note:** this plugin does not need to be explicitly loaded if the failover connection plugin is loaded.                                                                                                                                                                                                                                                                                                                                                |
```
Comment (brunos-bq) [inline docs/using-the-jdbc-driver/UsingTheJdbcDriver.md:111]: Was there a chance previously of someone not using the right dialect but then using this plugin for the Host list?

File: wrapper/src/main/java/software/amazon/jdbc/wrapper/ConnectionWrapper.java [hunk 5]
```diff
@@ -133,13 +150,16 @@ protected void init(
     if (this.pluginService.getCurrentConnection() == null) {
       final Connection conn =
           this.pluginManager.connect(
-              this.targetDriverProtocol, this.pluginService.getInitialConnectionHostSpec(), props, true);
+              this.targetDriverProtocol, this.pluginService.getInitialConnectionHostSpec(), props,
```
Comment (brunos-bq) [inline wrapper/src/main/java/software/amazon/jdbc/wrapper/ConnectionWrapper.java:153]: did anything change with line limits?

**Review by brunos-bq — COMMENTED** [review]
Checkstyle changes with feature changes reduce a bit of the readability

File: docs/using-the-jdbc-driver/UsingTheJdbcDriver.md (hunk 4 above)
Comment (karenc-bq) [reply docs/using-the-jdbc-driver/UsingTheJdbcDriver.md:111]: Unlikely but may happen

File: wrapper/src/main/java/software/amazon/jdbc/wrapper/ConnectionWrapper.java (hunk 1 above)
Comment (karenc-bq) [reply wrapper/src/main/java/software/amazon/jdbc/wrapper/ConnectionWrapper.java:114]: I think my editor was using the wrong line limit setting and I never noticed.

File: docs/using-the-jdbc-driver/UsingTheJdbcDriver.md (hunk 4 above)
Comment (brunos-bq) [reply docs/using-the-jdbc-driver/UsingTheJdbcDriver.md:111]: is it something we want to now make impossible?

Comment (karenc-bq) [reply docs/using-the-jdbc-driver/UsingTheJdbcDriver.md:111]: Could you please clarify?

File: .editorconfig [hunk 6]
```diff
@@ -21,7 +21,7 @@ ij_kotlin_code_style_defaults = KOTLIN_OFFICIAL
 ij_kotlin_name_count_to_use_star_import = 99
```
Comment (brunos-bq) [inline .editorconfig:21]: NIT: is this file supposed to be in the repo?

File: docs/using-the-jdbc-driver/UsingTheJdbcDriver.md (hunk 4 above)
Comment (brunos-bq) [reply docs/using-the-jdbc-driver/UsingTheJdbcDriver.md:111]: I was just wondering if there were people using the plugin without configuring the dialect properly.
If that was the case, this change will break their workflow. Hence the question, did we want that to be possible, or we should now force them to use the correct dialect.
If people were using the plugin with the wrong dialect, we need to ensure the documentation is there for them to fix their workflow.

Comment (sergiyvamz) [inline docs/using-the-jdbc-driver/UsingTheJdbcDriver.md:111]: Does it make sense to mark this plugin as deprecated? We can remove this deprecated plugin in 3-6 months. That would be enough time for users to adjust their code base and app configurations.

File: wrapper/src/main/java/software/amazon/jdbc/wrapper/ConnectionWrapper.java [hunk 7]
```diff
@@ -122,8 +124,19 @@ protected void init(
 
     this.pluginManager.init(this.pluginService, props, pluginManagerService);
 
-    this.hostListProviderService.setHostListProvider(
-        new ConnectionStringHostListProvider(props, this.originalUrl, this.hostListProviderService));
+    if (this.hostListProviderService.getDialect() instanceof TopologyAwareDatabaseCluster) {
```
Comment (sergiyvamz) [inline wrapper/src/main/java/software/amazon/jdbc/wrapper/ConnectionWrapper.java:127]: I'd keep ConnectionStringHostListProvider at this place.

File: wrapper/src/main/java/software/amazon/jdbc/wrapper/ConnectionWrapper.java [hunk 8]
```diff
@@ -122,8 +124,19 @@ protected void init(
...
+          this.originalUrl)
+      );
+    } else {
+      this.hostListProviderService.setHostListProvider(new ConnectionStringHostListProvider(
+          props,
+          this.originalUrl,
+          this.hostListProviderService));
+    }
 
     this.pluginManager.initHostProvider(
```
Comment (sergiyvamz) [inline wrapper/src/main/java/software/amazon/jdbc/wrapper/ConnectionWrapper.java:141]: After allowing plugins to set up their own host list provider, it's a perfect time to set up a host list provider mentioned in dialect (if any).

File: .editorconfig (hunk 6 above)
Comment (karenc-bq) [reply .editorconfig:21]: Yes

File: docs/development-guide/Pipelines.md [hunk 9]
```diff
@@ -67,7 +67,7 @@ it does not perform any additional work.
 
 For cases where keeping updated information on existing and available replicas is necessary,
 such as during the failover procedure, it is important to have a host list provider that can re-fetch information once in a while,
-like the [Aurora host list provider](/wrapper/src/main/java/software/amazon/jdbc/plugin/AuroraHostListConnectionPlugin.java).
+like the Aurora host list provider.
```
Comment (brunos-bq) [inline docs/development-guide/Pipelines.md:70]: NIT: should we mention that this aurora host list provider is available using the right dialect?

File: docs/using-the-jdbc-driver/UsingTheJdbcDriver.md [hunk 10]
```diff
@@ -120,6 +119,9 @@ The AWS JDBC Driver has several built-in plugins that are available to use. Plea
...
 | [Developer Plugin](./using-plugins/UsingTheDeveloperPlugin.md)                                 | `dev`                     | Any database           | Helps developers test various everyday scenarios including rare events like network outages and database cluster failover. The plugin allows injecting and raising an expected exception, then verifying how applications handle it.                                                                                                                                                                                                                                                                                   |
 
+:exclamation: **NOTE**: Aurora Host List Connection Plugin is deprecated. The wrapper is able to automatically set the Aurora host list provider for connections to Aurora MySQL and Aurora PostgreSQL databases.
```
Comment (brunos-bq) [inline docs/using-the-jdbc-driver/UsingTheJdbcDriver.md:122]: Is it automatically or does the user needs to do something to get it? If so, we need to mention it here

Comment (karenc-bq) [reply docs/using-the-jdbc-driver/UsingTheJdbcDriver.md:122]: Users shouldn't need to do anything

File: docs/development-guide/Pipelines.md (hunk 9 above)
Comment (karenc-bq) [reply docs/development-guide/Pipelines.md:70]: Hmm I updated it to say that it is available when connecting to aurora dbs.

File: buildSrc/build.gradle.kts [hunk 11]
```diff
@@ -28,7 +28,7 @@ repositories {
 }
 
 dependencies {
-    implementation("com.igormaznitsa:jcp:7.0.5")
+    implementation("com.igormaznitsa:jcp:7.0.2")
```
Comment (sergiyvamz) [inline buildSrc/build.gradle.kts:31]: why to downgrade it?

File: wrapper/build.gradle.kts [hunk 12]
```diff
@@ -39,10 +39,10 @@ dependencies {
 
     testImplementation("org.junit.platform:junit-platform-commons:1.9.0")
     testImplementation("org.junit.platform:junit-platform-engine:1.9.3")
-    testImplementation("org.junit.platform:junit-platform-launcher:1.10.0")
-    testImplementation("org.junit.platform:junit-platform-suite-engine:1.10.0")
+    testImplementation("org.junit.platform:junit-platform-launcher:1.9.0")
```
Comment (sergiyvamz) [inline wrapper/build.gradle.kts:42]: why?

File: wrapper/src/main/java/software/amazon/jdbc/plugin/failover/FailoverConnectionPlugin.java [hunk 13]
```diff
@@ -207,15 +207,18 @@ public <T, E extends Exception> T execute(
...
+      final Properties properties,
       final HostListProviderService hostListProviderService,
       final JdbcCallable<Void, SQLException> initHostProviderFunc)
       throws SQLException {
     initHostProvider(
         initialUrl,
         hostListProviderService,
         initHostProviderFunc,
-        () -> new AuroraHostListProvider(driverProtocol, hostListProviderService, props, initialUrl),
+        () -> this.hostListProviderService
```
Comment (sergiyvamz) [inline wrapper/src/main/java/software/amazon/jdbc/plugin/failover/FailoverConnectionPlugin.java:218]: I don't think we need this piece anymore.

File: wrapper/src/main/java/software/amazon/jdbc/wrapper/ConnectionWrapper.java [hunk 14]
```diff
@@ -121,9 +123,11 @@ protected void init(
     this.pluginManagerService = pluginManagerService;
 
     this.pluginManager.init(this.pluginService, props, pluginManagerService);
-
-    this.hostListProviderService.setHostListProvider(
-        new ConnectionStringHostListProvider(props, this.originalUrl, this.hostListProviderService));
+    final HostListProviderSupplier supplier = this.pluginService.getDialect().getHostListProvider();
+    if (supplier != null) {
+      final HostListProvider provider = supplier.getProvider(props, this.originalUrl, hostListProviderService);
+      hostListProviderService.setHostListProvider(provider);
```
Comment (sergiyvamz) [inline wrapper/src/main/java/software/amazon/jdbc/wrapper/ConnectionWrapper.java:129]: Check provider for null?

File: buildSrc/build.gradle.kts (hunk 11 above)
Comment (karenc-bq) [reply buildSrc/build.gradle.kts:31]: Not intentional

File: wrapper/src/main/java/software/amazon/jdbc/wrapper/ConnectionWrapper.java (hunk 14 above)
Comment (karenc-bq) [reply wrapper/src/main/java/software/amazon/jdbc/wrapper/ConnectionWrapper.java:129]: It is not nullable

**Review by sergiyvamz — APPROVED** [review]

**Review by brunos-bq — APPROVED** [review]

**Pull request merged**